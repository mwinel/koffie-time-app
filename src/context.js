import React, { Component } from "react";
import { Posts } from "./data";

const PostContext = React.createContext();

class PostProvider extends Component {
  state = {
    posts: [],
    notFeatured: [],
    featuredPosts: [],
    featuredTutorials: [],
    featuredEngineering: [],
    featuredCareers: [],
    featuredDesign: []
  };

  componentDidMount() {
    this.setPosts();
  }

  setPosts = () => {
    let tempPosts = [];
    let notFeatured = [];
    let featuredPosts = [];
    let featuredTutorials = [];
    let featuredEngineering = [];
    let featuredCareers = [];
    let featuredDesign = [];
    Posts.forEach(item => {
      const singlePost = { ...item };
      tempPosts = [...tempPosts, singlePost];
      // not featured.
      const notfeatured = tempPosts.filter(item => item.featured === false);
      notFeatured = [...notfeatured]
      // featured post.
      const featured = tempPosts.filter(item => item.featured === true);
      featuredPosts = [...featured.slice(0, 1)];
      // featured tutorials
      const tutorialCategory = "Tutorial";
      const tempTutorials = tempPosts.filter(
        item => item.category === tutorialCategory.toLowerCase()
      );
      featuredTutorials = [...tempTutorials.slice(0, 3)];
      // featured engineering
      const engineeringCategory = "Engineering";
      const tempEngineering = tempPosts.filter(
        item => item.category === engineeringCategory.toLowerCase()
      );
      featuredEngineering = [...tempEngineering.slice(0, 3)];
      // featured careers
      const careersCategory = "Career";
      const tempCareers = tempPosts.filter(
        item => item.category === careersCategory.toLowerCase()
      );
      featuredCareers = [...tempCareers.slice(0, 3)];
      // featured design
      const designCategory = "Design";
      const tempDesign = tempPosts.filter(
        item => item.category === designCategory.toLowerCase()
      );
      featuredDesign = [...tempDesign.slice(0, 3)];
    });
    this.setState(() => {
      return {
        posts: tempPosts,
        notFeatured: notFeatured.filter(item => item.featured === false),
        featuredPosts: featuredPosts,
        featuredTutorials: featuredTutorials.filter(item => item.featured === false),
        featuredEngineering: featuredEngineering.filter(item => item.featured === false),
        featuredCareers: featuredCareers.filter(item => item.featured === false),
        featuredDesign: featuredDesign.filter(item => item.featured === false)
      };
    });
  };

  render() {
    return (
      <PostContext.Provider value={{ ...this.state }}>
        {this.props.children}
      </PostContext.Provider>
    );
  }
}

const PostConsumer = PostContext.Consumer;

export { PostProvider, PostConsumer };
