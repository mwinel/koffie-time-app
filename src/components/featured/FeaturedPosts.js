import React, { Component } from "react";
import FeaturedPost from "./FeaturedPost";
import { PostConsumer } from "../../context";

class FeaturedPosts extends Component {
  render() {
    const bg = process.env.PUBLIC_URL + "/images/cool-background.png";
    const styles = {
      backgroundImage: `url(${bg})`,
      backgroundPosition: "0% 100%"
    };
    return (
      <section className="featured-section pb-4" style={styles}>
        <div className="container">
          <h6 className="grid-title pt-4 pb-2">Featured</h6>
          <div className="row">
            <PostConsumer>
              {value => {
                return value.featuredPosts.map(post => {
                  return <FeaturedPost key={post.id} post={post} />;
                });
              }}
            </PostConsumer>
          </div>
        </div>
      </section>
    );
  }
}

export default FeaturedPosts;
