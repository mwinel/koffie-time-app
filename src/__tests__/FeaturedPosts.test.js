import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { BrowserRouter } from "react-router-dom";
import pretty from "pretty";
import FeaturedPosts from "../components/featured/FeaturedPosts";
import { PostProvider } from "../context";
import value from "../testHelpers";

let container = null;

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it("renders featured-posts component", () => {
  act(() => {
    render(
      <BrowserRouter>
        <PostProvider>
          <FeaturedPosts {...value} />
        </PostProvider>
      </BrowserRouter>,
      container
    );
  });
  expect(container.innerHTML).toBeDefined();
  expect(pretty(container.innerHTML)).toMatchSnapshot();
});
