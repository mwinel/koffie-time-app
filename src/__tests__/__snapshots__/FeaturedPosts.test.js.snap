// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders featured-posts component 1`] = `
"<section class=\\"featured-section pb-4\\" style=\\"background-image: url(/images/cool-background.png); background-position: 0% 100%;\\">
  <div class=\\"container\\">
    <h6 class=\\"grid-title pt-4 pb-2\\">Featured</h6>
    <div class=\\"row\\">
      <div class=\\"col-sm-7\\"><a href=\\"/post\\">
          <div class=\\"featured-post-image-wrapper pb-5\\" style=\\"background-size: cover; background-repeat: no-repeat; background-position: 50% 50%; height: 350px;\\"></div>
        </a></div>
      <div class=\\"col-sm-5\\">
        <div class=\\"pb-2\\"><small>Jan 26 2018 </small> <span class=\\"dot-divider\\">Â·</span> <small>7 min read</small></div><a href=\\"/post\\">
          <h2 class=\\"card-title font-weight-bold\\">Setting clear career goals as a Software Engineer</h2>
        </a>
        <div class=\\"flex flex-row flex-nowrap\\">
          <div class=\\"py-2\\">
            <div class=\\"author-avatar mr-2\\" style=\\"background-size: cover; background-repeat: no-repeat; background-position: 50% 50%; width: 36px; height: 36px;\\"></div><small class=\\"username\\"><a href=\\"/\\">@davidblair</a></small>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>"
`;
